### PR TITLE
Custom containerd client for use in init.

### PR DESCRIPTION
**Pull request by ijc**
**- What I did**

Latest `ctr` has removed the `--runtime-config` option which we rely on. `ctr` is not (considered by containerd devs) to be a supported interface (it's more of a debug tool).

This PR replaces `/etc/init.d/010-containerd`'s use of `ctr` with a simple custom client written against the containerd client library.

**- How I did it**

The interesting commit message is:

>    init: replace ctr with a custom client using the containerd client library
>    
>    Currently it supports only `service start <SERVICE>`, but it could grow e.g.
>    `stop`, `exec` etc in the future (although you can still use `ctr` for those).
>    
>    I wasn't able yet to figure out how to make containerd do the rw bind step:
>        /bin/mount --bind "$f/rootfs" "$f/rootfs"
>        mount -o remount,rw "$f/rootfs"
>    So that remains for now.
>    
>    In order to be able to use go-compile.sh the containerd build needs to move
>    from /root/go to /go as the GOPATH.
>    
>    The vendoring situation is not ideal, but since this tool wants to be an exact
>    match for the containerd it seems tollerable to reuse its vendoring.
>    
>    Signed-off-by: Ian Campbell <ian.campbell@docker.com>

**- How to verify it**

Everything should just continue working.

**- Description for the changelog**

Switched away from `ctr` usage in init.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by ijc** [comment]
Forgot to mention that during the build I see:
```
go build...
# cmd/service
warning: unable to find runtime/cgo.a
```

But it doesn't appear to cause any harm

File: pkg/containerd/cmd/service/skanky-vendor.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
```
Comment (rn) [inline pkg/containerd/cmd/service/skanky-vendor.sh]: "does exactly what it says on the tin"

File: pkg/containerd/cmd/service/start.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/services/containers"
+	"github.com/containerd/containerd/api/services/execution"
+	//"github.com/containerd/containerd/api/types/mount"
```
Comment (rn) [inline pkg/containerd/cmd/service/start.go]: could you add a one line comment as to why this is commented out (or remove it)

Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: Oops, will remove.

File: pkg/containerd/cmd/service/start.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		enc.SetIndent("", "    ")
+		if err := enc.Encode(&spec); err != nil {
+			log.Fatalf("dumping spec: %s", err)
+		}
+
+	}
+
+	ctr, err := client.NewContainer(ctx, service, containerd.WithSpec(spec),
+		func(ctx context.Context, client *containerd.Client, c *containers.Container) error {
+			c.RootFS = service // This seems odd, but mimics ctr.
```
Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: A bit more experimentation suggests that in this mode of operation not initialising `c.RootFS` is fine (and correct).

File: pkg/containerd/cmd/service/start.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	logfile := filepath.Join("/var/log", service+".log")
+	// We just need this to exist.
+	if err := ioutil.WriteFile(logfile, []byte{}, 0666); err != nil {
+		log.Fatalf("touch logfile: %s", err)
+	}
+
+	tasks := client.TaskService()
+	rsp, err := tasks.Create(ctx, &execution.CreateRequest{
```
Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: This ought to be `ctr.NewTask` I think.

File: pkg/containerd/cmd/service/start.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	service := args[0]
+	rootfs := filepath.Join("/containers/services", service, "rootfs")
+	log.Infof("Starting service: %q", service)
+	log := log.WithFields(log.Fields{
+		"service": service,
+	})
+
+	client, err := containerd.New(*sock)
+	if err != nil {
+		log.Fatal("creating containerd client")
```
Comment (rn) [inline pkg/containerd/cmd/service/start.go]: should this log the `err`?

File: pkg/containerd/cmd/service/start.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	ctx := namespaces.WithNamespace(context.Background(), "default")
+
+	var spec *specs.Spec
+	specf, err := os.Open(filepath.Join("/containers/services", service, "config.json"))
+	if err != nil {
+		log.Fatalf("failed to read config.json: %s", err)
```
Comment (rn) [inline pkg/containerd/cmd/service/start.go]: should this be `%v` instead of `%s` (as well as in the other places below)?

File: pkg/containerd/cmd/service/start.go (hunk 5 above)
Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: Yes. Will fix.

File: pkg/containerd/cmd/service/start.go (hunk 6 above)
Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: No, `%s` is correct and causes `err.String()` to be called, `%v` prints out the raw underlying datatype which is usually not what you want.

Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: ...at least I know who is responsible for all the use of `"%v", err` in the code base ;-)

Comment (rn) [inline pkg/containerd/cmd/service/start.go]: that would be @justincormack not me. He made me convert some of them back to `%v` and seems to have a strong opinion about it...

Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: I stand corrected, I made `stdin` be `/dev/nullx` and with both `%s` and `%v` the output is:
```
FATA[0000] failed to create task: rpc error: code = NotFound desc = stat /dev/nullx: no such file or directory  service=getty
```
Closer reading of https://golang.org/pkg/fmt/ confirms that both `%v` and `%s` will, if the argument implements the `Error` interface, call the `Error()` method (which is what I meant when I said `String()` above):
>  If the format (which is implicitly %v for Println etc.) is valid for a string (%s %q %v %x %X), the following two rules apply:
>
> 4. If an operand implements the error interface, the Error method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any).
>
> 5. If an operand implements method String() string, that method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any). 

I was sure I had seen garbled and unhelpful error messages from the use of `%v` on error types in the past, I must have been confused.

File: pkg/containerd/cmd/service/skanky-vendor.sh [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+# and we conveniently have a checkout already. We actually prefer to
+# reuse containerd's vendoring for consistency anyway.
+
+set -eu
+ctrd=$1
+mkdir -p vendor/github.com/containerd/containerd
+ln -nfs $ctrd/* vendor/github.com/containerd/containerd
+rm -f vendor/github.com/containerd/containerd/vendor
+
+for p in github.com/Microsoft/go-winio \
```
Comment (ijc) [inline pkg/containerd/cmd/service/skanky-vendor.sh]: I don't think it should be any more difficult to maintain than a `vendor.conf`, either as the client code here changes or as the `CONTAINERD_COMMIT=` in the `Dockerfile` changes -- which is roughly equivalent in this context to changing the containerd entry in a `vendor.conf` and then syncing all the other entries from containerd's `vendor.conf`, which is how I would do things if I was doing this with a `vendor.conf`.

File: pkg/containerd/cmd/service/start.go (hunk 6 above)
Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: I wrote this code when I was still using the default `log` package but have subsequently switched to logrus. Probably the right thing to do is switch to `log.WithError(err).Fatal("blah blah")`.

**Comment by ijc** [comment]
Force pushed addressing all the issues, plus moved the remount of rootfs into the tool, dropped a now unnecessary `wait` from the initscript (not sure if it was ever necessary) and bumped to a newer (master from yesterday) containerd (which was previously blocked by the removal of `--runtime-config`).

I did try and go to today's containerd master, but it seems to be broken, will debug, but no need to block this PR I think.

**Comment by ijc** [comment]
@riyazdf 
> do you have a link to the PR that removed -runtime-config? I'm curious what the rationale was

It went away in https://github.com/containerd/containerd/pull/962, but not explicitly mentioned (not a supported tool, etcetc). IIRC it's not the first time it went away.

**Comment by ijc** [comment]
> I did try and go to today's containerd master, but it seems to be broken, will debug, but no need to block this PR I think.

I figured it out (`config.toml` needed an update) and have added today's containerd to the end of the series.

**Comment by justincormack** [comment]
@ijc the `runtime/cgo.a` is normal, must find out why eventually.

You have a rebase conflict with Rolf's init changes

File: pkg/containerd/cmd/service/skanky-vendor.sh (hunk 7 above)
Comment (justincormack) [inline pkg/containerd/cmd/service/skanky-vendor.sh]: Why can't we just copy all of `containerd`s `vendor` directory as we have a checkout? Having more things does not matter.

**Comment by ijc** [comment]
It's also going to conflict with #2044. I may as well wait for that (trivial) PR to be merged before rebasing

File: pkg/containerd/cmd/service/skanky-vendor.sh (hunk 7 above)
Comment (ijc) [inline pkg/containerd/cmd/service/skanky-vendor.sh]: We need containerd itself too, so it's not a simple `cp` of `vendor` but it might be simpler than this scheme.

File: pkg/containerd/cmd/service/start.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+			Source:      rootfs,
+			Options:     []string{"bind"},
+		},
+		{
+			Destination: "/",
+			Type:        "",
+			Source:      "",
+			Options:     []string{"remount", "rw"},
+		},
+	}, spec.Mounts...)
```
Comment (justincormack) [inline pkg/containerd/cmd/service/start.go]: I really don't like this, would rather do the remounts up front, not add them to the spec. These won't be needed if we use the image store, and they are not really part of the spec which should reflect what the user specified only, they are part of our implementation. Also they are only needed due to `runc` bugs.

Comment (justincormack) [inline pkg/containerd/cmd/service/start.go]: (you can leave the remount in the shell script for now if you like, I will probably replace it all with a Go wrapper soon anyway)

Comment (ijc) [inline pkg/containerd/cmd/service/start.go]: will do this tomorrow.

File: pkg/containerd/cmd/service/skanky-vendor.sh (hunk 7 above)
Comment (justincormack) [inline pkg/containerd/cmd/service/skanky-vendor.sh]: sure it needs containerd (ex its vendor too), I but it avoids listing them

File: pkg/getty/Dockerfile [hunk 9]
```diff
@@ -17,4 +17,4 @@ COPY --from=mirror /out/ /
 COPY usr/ /usr/
 COPY etc/ /etc/
 CMD ["/usr/bin/rungetty.sh"]
-LABEL org.mobyproject.config='{"pid": "host", "net":"host", "binds": ["/run:/run", "/etc:/hostroot/etc","/tmp/ctr:/tmp/ctr", "/usr/bin/ctr:/usr/bin/ctr", "/usr/bin/runc:/usr/bin/runc", "/usr/bin/dist:/usr/bin/dist", "/var:/var","/containers:/containers","/dev:/dev","/sys:/sys"], "capabilities": ["all"]}'
+LABEL org.mobyproject.config='{"pid": "host", "net":"host", "binds": ["/run:/run", "/etc:/hostroot/etc", "/usr/bin/ctr:/usr/bin/ctr", "/usr/bin/runc:/usr/bin/runc", "/usr/bin/dist:/usr/bin/dist", "/var:/var","/containers:/containers","/dev:/dev","/sys:/sys"], "capabilities": ["all"]}'
```
Comment (justincormack) [inline pkg/getty/Dockerfile]: Why did you remove `/tmp/ctr:/tmp/ctr`? You still need this if you want to run `ctr` from the getty

Comment (ijc) [inline pkg/getty/Dockerfile]: `/tmp/ctr` a) is not created is you aren't using `ctr` (which is the purpose of the tool) and b) isn't created even by newer versions of `ctr`.

The only thing it is/was used for AFAICT is to hold the stdio fifos from `ctr run` and `ctr exec`. There is/was nothing in there which is useful to any independent future run of `ctr`, you can run `ctr` in the `getty` container just fine without sharing `/tmp/ctr` with the host mount namespace.

**Comment by ijc** [comment]
Force pushed with all comments (apart from the `/tmp/ctr` thing, still being discussed above) fixed.

File: pkg/getty/Dockerfile (hunk 9 above)
Comment (justincormack) [inline pkg/getty/Dockerfile]: ok
